src/libpython.cpp

```diff
@@ -453,39 +457,57 @@ bool import_numpy_api(bool python3, std::string* pError) {
   return true;
 }
 
+// returns 'true' if the buffer was flushed, or if the stdout / stderr
+// objects within 'sys' did not contain 'flush' methods
+bool flush_std_buffer(const char* name) {
 
-int flush_std_buffers() {
-  int status = 0;
-  PyObject* tmp = NULL;
-  PyObject *error_type, *error_value, *error_traceback;
-  PyErr_Fetch(&error_type, &error_value, &error_traceback);
+  // returns borrowed reference
+  PyObject* buffer(PySys_GetObject(name));
+  if (buffer == NULL || buffer == Py_None)
+    return true;
 
-  PyObject* sys_stdout(PySys_GetObject("stdout"));  // returns borrowed reference
-  if (sys_stdout == NULL)
-    status = -1;
-  else
-    tmp = PyObject_CallMethod(sys_stdout, "flush", NULL);
+  // try to invoke flush method
+  if (!PyObject_HasAttrString(buffer, "flush"))
```

Comment:
This probably doesn't matter, but I noticed in the python sources that `PyObject_HasAttr()` just calls `PyObject_GetAttr()` and then checks if an exception was raised. It's trivial, but we're on a hot code path here and the approach seems unnecessarily expensive if we're going to immedeatly fetch the attr again anyway in the common case.

More relevant however, is that `flush` should always be present.
`sys.stdout` and `sys.stderr` are supposed to be a file object, and the abstract io base class defines this mixin method always: https://docs.python.org/3/library/io.html#io.IOBase (the method does nothing for output streams where
flushing doesn't apply)

If `flush` missing, I think we should emit some error output, because it indicates that something is "wrong".

Reply:
FWIW we have this code that apparently expects stdout to be null in some scenarios.

https://github.com/rstudio/reticulate/blob/bugfix/windows-interrupts/R/output.R#L9-L13